segmentation.py

```diff
@@ -0,0 +1,575 @@
+            return
+
+        # Sort the hypotheses in decreasing log probability order.
+        self.__hypotheses.sort(key=lambda dct: -dct['log_probability'])
+
+        # Store the indices of likely hypotheses.
+        minprob = np.log(minprob)
+        index = [i for i, hypot in enumerate(self.__hypotheses)
+                 if hypot['log_probability'] > minprob]
+
+        # Trim the hypotheses.
+        index = index[:maxhypot] if len(index) >= maxhypot else index
+        self.__hypotheses = [self.__hypotheses[i] for i in index]
+
+        # NOTE: This final ordering can preserve the original order of the
+        #       hypotheses. Interestingly, the algorithm specified in update
+        #       does not require that the hypotheses be ordered! This sort can
+        #       safely be ignored.
+        # self.__hypotheses.sort(key=lambda dct: dct['index'])
+
+        # Normalize the hypotheses so that their probabilities sum to one.
+        logsum = -np.inf
+        for hypot in self.__hypotheses:
+            logsum = self.__accum(logsum, hypot['log_probability'])
+        for hypot in self.__hypotheses:
+            hypot['log_probability'] -= logsum
+
+    def segment(self):
+        """Return posterior probabilities OR sequence segmentation.
+
+        If the MAX-PRODUCT algorithm is selected, this method returns the most
+        likely sequence segmentation as a list of integers. Each integer in the
+        list marks where a segment begins.
+
+        If the SUM-PRODUCT algorithm is selected, this method returns the
+        posterior probabilities of the segmentation hypotheses as a numpy
+        array. Rows in the array represent hypotheses and columns in the array
+        represent data points in the time-series.
+
+        Returns:
+            object: If the MAX-PRODUCT algorithm is selected a list containing
+                    the most likely sequence segmentation is returned. If the
+                    SUM-PRODUCT algorithm is selected, a numpy array containing
+                    the posterior probabilities of the segmentation hypotheses
+                    is returned.
+
+        """
+
+        # In the max-product algorithm, the most likely hypotheses are
+        # tracked. Recover the most likely segment boundaries by performing a
+        # back-trace.
+        if self.__alg__ == 'maxprod':
+
+            # Find the most likely hypothesis.
+            max_hypothesis = max(self.__hypotheses,
+                                 key=lambda dct: dct['log_probability'])
+
+            # Find the best sequence segmentation given all the data so far.
+            segment_boundaries = [len(self.__index) - 1, ]
+            index = segment_boundaries[0] - 1
+            count = max_hypothesis['count'] - 1
+            while index > 0:
+                index -= count
+                segment_boundaries.insert(0, index)
+                count = self.__index[index - 1]
+
+            return segment_boundaries
```

Comment:
Actually I modified this a little. I changed the code to add segment boundaries to the head of the list (rather than the tail) so that segment_boundaries.reverse() would not need to be called. Other than that I believe it is the same.